records/2023-01-24.malicious-content-scanning.md

```diff
@@ -0,0 +1,104 @@
+
+We have explored using the `asyncio` library in python to create a new process to call the scan-files api [here](https://github.com/cds-snc/notification-document-download-api/pull/97/files#diff-a842b15bd72bea41437e64bda92d12262c8eca35af7e8adf111bd520fe7c3010R1). Then once the result is returned, we could write that result to the S3 object as a tag which can be read later. However, this approach did not work to reduce the response time for the Document Download API.
+
+It seems that it is [not possible to create background processes inside flask](https://flask.palletsprojects.com/en/2.2.x/async-await/#background-tasks).
+
+## Options
+
+Here are some of the options for scanning email attachments with the scan-files project.
+
+### Duplicate S3 bucket
+
+The scan-files project comes with a terraform module that triggers a scan of newly created S3 objects in a bucket it is told to watch. The S3 bucket we are currently using to store file attachments won't work, since these file are each encrypted with a unique key and the scan-files tf module does not support this. So, we could create a 2nd S3 bucket with restricted access containing duplicates of these files, all encrypted with the same key.
+
+Mapping files from one S3 bucket to the other is simple - we can just use the same file name in both cases. We would need to enforce retention/deletion policies in the duplicate bucket that are at least as strict. And the duplicate S3 bucket can have an access policy that restricts access to the document-download API and scan-files.
+
+We only need to store the duplicate files until the malware scan is complete and the scan verdict is read by notification-api prior to sending the email and file attachment (either in linked or attached directly).
+
+Here is what happens when an API user makes a request to send an email with an attachment:
+
+```mermaid
+flowchart TD
+    A[user makes request to /email with file attachment]
+    B[notification-api uploads file to dd-api]
+    C[dd-api uploads file to:\n document-download S3 bucket \n scan-files S3 bucket]
+    D[dd-api returns a 201 to notification-api]
+    E[notification-api returns a 201 to user, sets notification status to 'Sending']
+    A-->B
+    B-->C
+    C-->D
+    D-->E
+```
+Eventually the email sending job is processed by Celery, and the following occurs: 
+```mermaid
+flowchart TD
+    A[Celery calls `send_email_to_provider` fn]
+    B{Does email contain a file attachment?}
+    C1{request to dd-api to \n/check-scan-verdict endpoint}
+    C2[Send email as usual]
+    D[Wait for 5 minutes and try again]
+    E[Send email]
+    F[Do not send email\n Set notification status to 'malware-scan-failure']
```

Comment:
Out of curiosity, how is this `malware-scan-failure` communicated back to the user who tried to send the email with the malicious attachment?

Also, it's possible that a few different [scan verdicts](https://github.com/cds-snc/scan-files/blob/233cca8d3f400d35616d76b71479ddf0238d7554/api/models/Scan.py#L16-L23) could be returned:
```
IN_PROGRESS = "in_progress"
CLEAN = "clean"
MALICIOUS = "malicious"
ERROR = "error"
UNABLE_TO_SCAN = "unable_to_scan"
```

Reply:
For the statuses "malicious" and "suspicious":

My thinking is that the email would be marked as "Failed" and there would be an explanation like "Malware scan failure" - we don't have the exact content yet. That message would appear as the red text in the image below:
<img width="1054" alt="Screen Shot 2023-02-07 at 11 08 55 AM" src="https://user-images.githubusercontent.com/5498428/217300910-387ad63f-8b12-4721-9bed-918b3ca66912.png">

For "in_progress": 
Try again in 5 minutes.

For all other statuses:
I think we would just send the email as usual so that we don't block sending if scan-files goes down or there is a weird error. But maybe we could have an alert to tell us if we are getting a lot of errors.

Reply:
Ah that makes sense 👍🏻

Reply:
>I think we would just send the email as usual so that we don't block sending if scan-files goes down or there is a weird error. But maybe we could have an alert to tell us if we are getting a lot of errors.

+1 for raising an alarm if we get many errors.

Can we please add more paths to the decision tree diagram in case that malware scanning returns an error, or is down? to demonstrate case when we actually send the emails even in the case of uncertainty.

Reply:
I added the error case and also the case where the scan is taking too long because `scan-files` is down.